core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -63,6 +65,8 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
         doWork()
       }
     } catch{
+      case fatalExitError: FatalExitError =>
+        fatalExitError.shutdownSystem()
```

Comment:
This aproach does not work because shutdown hook waits for ShutdownableThread and vice versa.
I suggest you to make a global thread that handles unexepected exceptions:
```
object KafkaServer {
...
  val unexpectedQueue = new java.util.concurrent.LinkedBlockingQueue[Throwable]()
  val unexpectedThread = new Thread(new Runnable {
    import KafkaServer._
    override def run(): Unit = {
      unexpectedQueue.take() match {
        case e: FatalExitError =>
          e.shutdownSystem()
        case _ =>
          System.exit(1)
      }
    }
  })
  unexpectedThread.start()
...
```
And instead of doing fatalExitError.shutdownSystem() here just execute unexpectedQueue.put(fatalExitError)
This approach works for me. I've just checked it :)

cc @nepal